src/apps/morpho/ethereum/morpho.morpho-blue.contract-position-fetcher.ts

```diff
@@ -0,0 +1,210 @@
+    @Inject(MorphoViemContractFactory) protected readonly contractFactory: MorphoViemContractFactory,
+  ) {
+    super(appToolkit);
+  }
+
+  getContract(address: string) {
+    return this.contractFactory.morphoBlue({ address, network: this.network });
+  }
+
+  async _fetchCollateralData(): Promise<MarketCollateralResponse> {
+    return gqlFetch<MarketCollateralResponse>({
+      endpoint: this.subgraphUrl,
+      query: MorphoBlueMarkets.MARKET_COLLATERAL_QUERY,
+      variables: { ids: MorphoBlueMarkets.whitelistedIds },
+    });
+  }
+
+  async getDefinitions({ multicall }: GetDefinitionsParams) {
+    const morphoBlue = this.contractFactory.morphoBlue({ address: this.morphoAddress, network: this.network });
+    const morpho = multicall.wrap(morphoBlue);
+    const marketData = await Promise.all(
+      this.whitelistedIds.map(async id => {
+        const [loanToken, collateralToken, oracle, irm, lltv] = await morpho.read.idToMarketParams([id]);
+        return { loanToken, collateralToken, oracle, irm, lltv };
+      }),
+    );
+    return marketData.map((data, i) => ({
+      address: this.morphoAddress,
+      marketId: this.whitelistedIds[i].toLowerCase(),
+      loanTokenAddress: data.loanToken.toLowerCase(),
+      collateralTokenAddress: data.collateralToken.toLowerCase(),
+      irmAddress: data.irm.toLowerCase(),
+      oracleAddress: data.oracle.toLowerCase(),
+      lltv: data.lltv,
+    }));
+  }
+
+  async getDataProps({
+    contractPosition,
+    multicall,
+    definition: { marketId, loanTokenAddress, collateralTokenAddress, oracleAddress, irmAddress, lltv },
+  }: GetDataPropsParams<MorphoBlue, MorphoContractPositionDataProps, MorphoContractPositionDefinition>) {
+    const morpho = multicall.wrap(
+      this.contractFactory.morphoBlue({ address: this.morphoAddress, network: this.network }),
+    );
+
+    const [totalSupplyAssets, totalSupplyShares, totalBorrowAssets, totalBorrowShares, lastUpdate, fee] =
+      await morpho.read.market([marketId]);
+    const irm = this.contractFactory.morphoAdaptiveCurve({ address: irmAddress, network: this.network });
+    const borrowRate = await irm.read.borrowRateView([
+      {
+        loanToken: loanTokenAddress,
+        collateralToken: collateralTokenAddress,
+        oracle: oracleAddress,
+        irm: irmAddress,
+        lltv: lltv,
+      },
+      {
+        totalSupplyAssets,
+        totalSupplyShares,
+        totalBorrowAssets,
+        totalBorrowShares,
+        lastUpdate,
+        fee,
+      },
+    ]);
+
+    const borrowRateYear = borrowRate * BigInt(SECONDS_PER_YEAR);
+    const borrowApy: number = +formatUnits(expN(borrowRateYear, BigInt(3), WAD, mulDivDown) - WAD, 18);
+    const utilization: bigint =
+      totalSupplyAssets != BigInt(0) ? wadDivDown(totalBorrowAssets, totalSupplyAssets) : BigInt(1);
+
+    const supplyRate = wadMulDown(wadMulDown(utilization, borrowRateYear), BigInt(WAD) - BigInt(fee));
```

Comment:
Beware this formula may not correspond to the actual experienced supply APY. This is still under discussion: https://github.com/morpho-org/morpho-blue-snippets/pull/45#issuecomment-1901879204

Reply:
I keep the implementation merged on the PR you mentionned, will update here if we decide otherwise: 
<img width="986" alt="image" src="https://github.com/morpho-labs/zapper-studio/assets/75627352/8328d987-8a40-4d79-8dc0-e01f7eaa80be">

Reply:
The implementation merged in the referenced PR does not correspond to the one of this PR ^^

ie:
```suggestion
    const supplyApy = wadMulDown(wadMulDown(utilization, borrowApy), BigInt(WAD) - BigInt(fee));
```

and the `supplyRate` should not exist!